impl/src/main/java/org/ehcache/impl/internal/classes/ClassInstanceProvider.java

```diff
@@ -49,6 +51,7 @@
    * Instances provided by this provider vs their counts.
    */
   protected final ConcurrentWeakIdentityHashMap<T, AtomicInteger> providedVsCount = new ConcurrentWeakIdentityHashMap<T, AtomicInteger>();
+  protected final Set<T> instantiated = new HashSet<T>();
```

Comment:
You probably want to use a `Set` with an reference identity contract and not an object identity one. See our `ConcurrentWeakIdentityHashMap` for example, which can easily be wrapped to act as a `Set` like we do in some `Store.Provider`.